### PR TITLE
Introduce special handling for 'deleted client'

### LINKED ISSUE #168: Can I catch and "reframe" error thrown when we disable an oauth app?
When we roll the tidyverse oauth app and actually disable the previous one, can I catch the error users will see if they haven't upgraded client package + gargle and supplement it with a probably diagnosis and suggested solution?

Obviously this won't help this time, but it could help next time.

### PR DESCRIPTION
**Pull request by jennybc**
Closes #168

### CODE REVIEW

File: R/Gargle-class.R [hunk 1]
```diff
@@ -208,12 +208,14 @@ Gargle2.0 <- R6::R6Class("Gargle2.0", inherit = httr::Token2.0, list(
   },
   #' @description (Attempt to) refresh a Gargle2.0 token
   refresh = function() {
-    cred <- get("refresh_oauth2.0", asNamespace("httr"))(
-      self$endpoint, self$app,
-      self$credentials, self$params$user_params, self$params$use_basic_auth
+    cred <- refresh_oauth2.0(
```
Comment (jennybc) [inline R/Gargle-class.R:211]: I was already doing something janky with this method in order to use a `Gargle2.0`-specific cache-removal procedure , so it feels fine/good to take the next step and fully implement `refresh_oauth2.0()` in gargle.

File: R/Gargle-class.R [hunk 2]
```diff
@@ -208,12 +208,14 @@ Gargle2.0 <- R6::R6Class("Gargle2.0", inherit = httr::Token2.0, list(
...
+      self$endpoint, self$app, self$credentials,
+      package = self$package
     )
     if (is.null(cred)) {
       token_remove_from_cache(self)
+      # TODO: return(NULL) here? why do we return the current, invalid,
+      # unrefreshed token?
```
Comment (jennybc) [inline R/Gargle-class.R:218]: The more I think about this, I think, when a token fails to refresh, we should clear its refresh token. And possibly do more to indicate that it's invalid and why?

Long-term, httr2-wise, I think a token needs a way to communicate its viability back to a client package that may be managing it in memory.

File: R/oauth-cache.R [hunk 3]
```diff
@@ -166,6 +166,8 @@ cache_clean <- function(cache, pattern = "-calliope$") {
   TRUE
 }
 
+gargle_legacy_app_pattern <- function() "-calliope$"
```
Comment (jennybc) [inline R/oauth-cache.R:169]: I assume this would evolve over time, as we roll the apps.

File: R/oauth-refresh.R [hunk 4]
```diff
@@ -0,0 +1,113 @@
+# this file has its origins in oauth-refresh.R and oauth-error.R in httr
+# I want to introduce behaviour to error informatively for a deleted OAuth app
+
+# Refresh an OAuth 2.0 credential.
+#
+# Refreshes the given token, and returns a new credential with a
+# valid access_token. Based on:
+# https://developers.google.com/identity/protocols/oauth2/native-app#offline
+refresh_oauth2.0 <- function(endpoint, app, credentials, package = NULL) {
```
Comment (jennybc) [inline R/oauth-refresh.R:9]: Pretty faithful copy of `httr::refresh_oauth2.0()`, but with some simplifications re: things that can't happen with Google auth + a relaxation of rigid lack-of-refresh error types.

File: R/oauth-refresh.R [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+# This implements error checking according to the OAuth2.0
+# specification: https://tools.ietf.org/html/rfc6749#section-5.2
+find_oauth2.0_error <- function(response) {
+  if (!httr::status_code(response) %in% oauth2.0_error_codes) {
+    return(NULL)
+  }
+
+  content <- httr::content(response)
+  if (is.null(content$error)) {
```
Comment (jennybc) [inline R/oauth-refresh.R:48]: This is more lax than httr, which checks against a fixed set of strings that, apparently, Google violates.

https://github.com/r-lib/httr/commit/ce8439f2d571e91bf846b43d5db3d89610568fa1

**Comment by jennybc** [comment]
I'm happy to receive feedback. But I'm also making and reviewing PR more as a way to document some analysis that might be useful for httr2. I'm punting on a few things here in a way that feels fairly harmless / way better than doing nothing.

File: R/Gargle-class.R (hunk 2 above)
Comment (hadley) [reply R/Gargle-class.R:218]: Yeah, seems like this should error as well as removing from the cache?

File: R/oauth-refresh.R (hunk 5 above)
Comment (hadley) [reply R/oauth-refresh.R:48]: That was probably written when I mistakenly believed that most APIs followed the OAuth spec.

**Review by hadley — APPROVED** [review]

**Pull request merged**